# Module 12, Linear Algebra for Machine Learning

**_Author: Carleton Smith_**

**Expected time = 2.5 hours**

**Total points = 120 points**



    
## Assignment Overview

This assignment will both test your understand of linear algebra and your ability to translate procedures into code. Some questions will ask you to create your own implementations of linear algebra operations, while others will allow you to take advantage of existing implementations available in libraries, such as `numpy` and `scipy`.
In this assignment you will implement basic and intermediate linear algebra operations in Python.You will begin by performing matrix addition, substraction, and multplication. You will also use dot product and cross product to perform calculations. You will then move on to more advanced operations such as transpotion, matrix determinant, and inverse. 

This assignment is designed to build your familiarity and comfort coding in Python while also helping you review key topics from each module. As you progress through the assignment, answers will get increasingly complex. It is important that you adopt a data scientist's mindset when completing this assignment. **Remember to run your code from each cell before submitting your assignment.** Running your code beforehand will notify you of errors and give you a chance to fix your errors before submitting. You should view your Vocareum submission as if you are delivering a final project to your manager or client. 

***Vocareum Tips***
- Do not add arguments or options to functions unless you are specifically asked to. This will cause an error in Vocareum.
- Do not use a library unless you are expicitly asked to in the question. 
- You can download the Grading Report after submitting the assignment. This will include feedback and hints on incorrect questions. 

### Learning Objectives

- Apply foundational linear algebra concepts in python to build basic algorithims.
- Use matrix operations to solve equations. 
- Perform advanced linear algebra procedures, such as PCA and Singular Value Decompositon (SVD)
- Apply linear algebra to solve least squares linear regression




## Index:

#### Module 12: Linear Algebra

##### Basic Matrix Operations
- [Question 1](#Question-1)
- [Question 2](#Question-2)
- [Question 3](#Question-3)
- [Question 4](#Question-4)
- [Question 5](#Question-5)

##### Intermediate Linear Algebra
- [Question 6](#Question-6)
- [Question 7](#Question-7)
- [Question 8](#Question-8)
- [Question 9](#Question-9)

##### Applied Linear Algebra
- [Question 10](#Question-10)
- [Question 11](#Question-11)
- [Question 12](#Question-12)



## Module 12: Linear Algebra

**Implement fundamental linear algebra procedures with Python**

This assignment will test your ability to demonstrate your understanding of fundamental linear algebra procedures and your ability to translate these procedures into code. Many (if not all) of the linear algebra procedures discussed in this assignment have existing implementations in Python. In practice, you will leverage these prebuilt implementations, but for the purpose of understanding the concepts and practicing Python, you will be asked to produce your own implementations.

The below functions are helper functions to assist with various tasks throughout the assignment. You can ignore them.

In [1]:
import numpy as np
import random
import inspect
import statsmodels.api as sm
from scipy.linalg import svd
from sklearn.datasets import make_regression
import matplotlib.pyplot as plt

In [2]:
def test_for_banned_funcs_and_oper(func, banned_funcs, banned_oper=[]):
    '''This function will test if your code uses a forbidden function'''
    unbound_closures = inspect.getclosurevars(func).unbound
    
    # check for banned operators
    student_code = inspect.getsourcelines(func)[0]
    if banned_oper:
        for line in student_code:
            for op in banned_oper:
                assert op not in line, "Cannot use operator(s) like '{}' in solution.".format(op)
    
    funcs_in_violation = []
    for uc in unbound_closures:
        if uc in banned_funcs:
            if uc == 'T':
                uc = '.T'
            funcs_in_violation.append(uc)
    assert len(funcs_in_violation) == 0,"Cannot use function(s) '{}' in solution.".format("' or '".join(funcs_in_violation))


In [3]:
def matrix_maker(*args, disp=True):
    '''This function will create an arbitrary number of matrices of specific shapes'''
    all_mats = []
    if not len(args) % 2:
        num_mats = int(len(args)/2)
        for _ in np.arange(num_mats): 
            n, m, *args = args
            if not m:
                M = np.random.randint(10, size=(n,))
            else:
                M = np.random.randint(10, size=(n,m))
            all_mats.append(M)
        if disp:
            for mat in all_mats:
                display(mat)
        if len(all_mats) == 1:
            return all_mats[0]
        return tuple(all_mats)
    else:
        raise ValueError("Must give even number of dimensions for each matrix.")



## Basic Matrix Operations
The following questions will ask you to create simple functions that perform basic matrix operations.

[Back to top](#Index:) 

### Question 1

*4 Points*

Your first task is to create a function called `matrix_adder` that will accept two matrices of the same shape as parameters and return another numpy array that is the result of matrix addition.

    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m) that is the sum of A + B.
    
    *If input matrices are of invalid shape, return 'Invalid dimensions'.*

Use the matrices `A` and `B` below as test cases.

In [4]:
A = np.array([
    [7, 2, 0, 0],
    [8, 8, 1, 9],
    [8, 6, 6, 8]
])

B = np.array([
    [7, 7, 8, 5],
    [5, 2, 5, 0],
    [0, 5, 5, 2]
])

In [5]:
### GRADED

### YOUR SOLUTION HERE
def matrix_adder(matrix1, matrix2):
    """Perform matrix addition with two input matrices.
    
    Parameters:
    matrix1 -- numpy array of shape (n,m)
    matrix2 -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m); the result of matrix addition.
    
    If input matrices are of invalid shape, return 'Invalid dimensions'.
    """
    pass

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*25)
print("Input Matrices:")
display(A, B)
print('-'*25)
print("Your Solution:")
display(matrix_adder(A, B))

# wrong dimensions
A2, B2 = matrix_maker(3,5,3,4, disp=False)
print('-'*25)
print("Invalid dimension case:\t{}".format(matrix_adder(A2, B2)))

-------------------------
Input Matrices:


array([[7, 2, 0, 0],
       [8, 8, 1, 9],
       [8, 6, 6, 8]])

array([[7, 7, 8, 5],
       [5, 2, 5, 0],
       [0, 5, 5, 2]])

-------------------------
Your Solution:


array([[14,  9,  8,  5],
       [13, 10,  6,  9],
       [ 8, 11, 11, 10]])

-------------------------
Invalid dimension case:	Invalid dimensions


In [6]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 2

*4 Points*

Similarly to Q1, create a function called `matrix_subtraction` that will accept two matrices of the same shape as parameters and return another numpy array that is the result of matrix subtraction.

    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m) that is the difference of A and B.

Use the matrices `A` and `B` below as test cases.

In [7]:
A = np.array([
    [0, 8, 6, 4],
    [2, 9, 0, 9],
    [1, 2, 1, 9]
])

B = np.array([
    [3, 4, 4, 3],
    [2, 9, 2, 9],
    [0, 2, 6, 5]
])

In [8]:
def matrix_subtraction(matrix1, matrix2):
    """Perform matrix subtraction with two input matrices.
    
    Parameters:
    matrix1 -- numpy array of shape (n,m)
    matrix2 -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (n,m); the result of matrix subtraction.
    
    If input matrices are of invalid shape, return 'Invalid dimensions'.
    """
    pass

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*25)
print("Input Matrices:")
display(A, B)
print('-'*25)
print("Your Solution:")
display(matrix_subtraction(A, B))

# wrong dimensions
A2, B2 = matrix_maker(3,5,3,4, disp=False)
print('-'*25)
print("Invalid dimension case:\t{}".format(matrix_subtraction(A2, B2)))

-------------------------
Input Matrices:


array([[0, 8, 6, 4],
       [2, 9, 0, 9],
       [1, 2, 1, 9]])

array([[3, 4, 4, 3],
       [2, 9, 2, 9],
       [0, 2, 6, 5]])

-------------------------
Your Solution:


array([[-3,  4,  2,  1],
       [ 0,  0, -2,  0],
       [ 1,  0, -5,  4]])

-------------------------
Invalid dimension case:	Invalid dimensions


In [9]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 3

*8 Points*

Your next task will be to implement another routine task in linear algebra - transpose a matrix. **For this exercise, you are not allowed to use any existing function or method to produce the transpose**. The reason for this limitation is to provide opportunities to practice manipulating `numpy` arrays natively.

Examples of fobidden functions:
    - numpy.transpose
    - numpy.matrix.transpose
    - numpy.ndarray.T
    - pandas.DataFrame.T
    - pandas.DataFrame.transpose

Create a function called `transpose` that will accept a single matrix (all numeric - int or float) of shape (n,m) as a parameter and return the transpose of that matrix as a numpy array of shape (m,n).

    Parameters:
    A -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (m,n) that is the transpose of A.

Use the matrices `A` below as a test case.

In [10]:
A = np.array([
    [6, 6, 0, 8],
    [7, 8, 3, 2],
    [7, 4, 8, 6]
])

In [11]:
def transpose(X):
    """Perform matrix transpose.
    
    Parameters:
    X -- numpy array of shape (n,m)
    
    Returns: a numpy array of shape (m,n); the transpose of matrix X.
    
    If X is 1 dimensional, return the original array.
    """
    pass

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*25)
print("Input Matrix:")
display(A)
print('-'*25)
print("Your Solution:")
display(transpose(A))

# one dimension array validation
A2 = matrix_maker(3,None,disp=False)
print('-'*25)
print("Input array:\t\t{}".format(A2))
print("One dimesional case:\t{}".format(transpose(A2)))

-------------------------
Input Matrix:


array([[6, 6, 0, 8],
       [7, 8, 3, 2],
       [7, 4, 8, 6]])

-------------------------
Your Solution:


array([[6., 7., 7.],
       [6., 8., 4.],
       [0., 3., 8.],
       [8., 2., 6.]])

-------------------------
Input array:		[8 1 8]
One dimesional case:	[8 1 8]


In [12]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 4

*9 Points*

Create a function called `dot_prod` that computes the dot product between two 1 dimensional vectors and return the result as an integer.

To compute the dot product, consider the two vectors below:

$$a = \begin{bmatrix} a_1 \\ \vdots \\ a_n \end{bmatrix}     b = \begin{bmatrix} b_1 \\ \vdots \\ b_n \end{bmatrix}$$

The dot product calculation is as follows:

$$a \cdot b = a_1b_1 + ... + a_nb_n$$

**For this exercise, you are not allowed to use any existing function or method to produce the dot product**. For the same reason mentioned in the earlier, this limitation is in place to enable coding practice, as well as reinforcing the mechanics of the linear algebra calculation.

Examples of fobidden functions:
    - numpy.dot()
    - numpy.inner()
    - numpy.matmul()
    - pandas.DataFrame.dot()
    - pandas.Series.dot()
    - the `@` operator

    Parameters:
    vecA -- a 1 dimensional numpy array of shape (n,)
    vecB -- a 1 dimensional numpy array of shape (n,)
    
    Returns: int; the dot product of the vectors vecA and vecB

Use the matrices `A` below as a test case.

In [13]:
a = np.array([3, 4, 2])
b = np.array([5, 5, 0])

In [14]:
### GRADED

### YOUR SOLUTION HERE
def dot_prod(vecA, vecB):
    """Calculate the dot product between two 1D vectors.
    
    Parameters:
    vecA -- a 1 dimensional numpy array of shape (n,)
    vecB -- a 1 dimensional numpy array of shape (n,)
    
    Returns: int or float; the dot product of the vectors vecA and vecB
    """
    pass

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*25)
print("Vector A:")
display(a)
print('-'*25)
print("Vector B:")
display(b)
print('-'*25)
print("Your Solution: {}".format(dot_prod(a,b)))
print("True Solution: {}".format(np.dot(a,b)))

-------------------------
Vector A:


array([3, 4, 2])

-------------------------
Vector B:


array([5, 5, 0])

-------------------------
Your Solution: 35
True Solution: 35


In [15]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 5

*10 Points*

Create a function that performs matrix multiplication between two 2 dimensional matrices. Your function should accept two matrices of appropriate dimensions and return a numpy array that is the result of matrix multiplication. See [this resource](https://www.mathsisfun.com/algebra/matrix-multiplying.html) for a simple tutorial.

    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (m,p)
    
    Returns: a numpy array of shape (n,p) that is the solution of AB matrix multiplication

**Your function should:**

1. Require `A` and `B` to have the acceptable dimensions for matrix multiplication.

    - If A or B have incompatible dimensions, your function should **return** the string `"Incorrect Dimensions"`

    
2. Avoid using the following functions:


    - numpy.dot()
    - numpy.inner()
    - numpy.matmul()
    - pandas.DataFrame.dot()
    - pandas.Series.dot()
    - the `@` operator


Use the example matrices below for testing.

In [16]:
A = np.array([
    [2,1],
    [1,2],
    [3,2]
])

B = np.array([
    [-1, 4],
    [1, -3]
])

In [17]:
### GRADED

### YOUR SOLUTION HERE
def matrix_mult(A, B):
    """Perform matrix multiplication with matrices A and B.
    
    Parameters:
    A -- numpy array of shape (n,m)
    B -- numpy array of shape (m,p)
    
    Returns:
    a numpy array of shape (n,p); the result of AB matrix multiplication
    """
    pass

###
### YOUR CODE HERE
###


In [18]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

## Intermediate Linear Algebra

### Question 6

*10 Points*

Create a function called `determinant2x2` that will calculate the determinant of any 2x2 matrix. You may use the 2x2 matrix, `A2`, below as a test case. See [this resource](https://www.mathsisfun.com/algebra/matrix-determinant.html) for a simple tutorial.

    
    Parameters:
    M -- numpy array of shape (n,n)
    
    Returns: float; the determinant of M.
    
For this exercise, avoid using the `numpy.linalg.det()` numpy function.

In [19]:
A2 = np.array([
    [5.5, 7],
    [2, 1]
])

In [20]:
### GRADED

### YOUR SOLUTION HERE
def determinant2x2(M):
    """Calculate the determinant of the 2x2 matrix, M.
    
    Parameters:
    M -- numpy array of shape (n,n)
    
    Returns: float; the determinant of M.
    """
    pass

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*20)
print("Sample Matrix:")
A2 = matrix_maker(2,2)
print('-'*20)
print("Numpy Solution:\t{}".format(np.linalg.det(A2)))
print("Your Solution:\t{}".format(determinant2x2(A2)))
print()
print('(Rounding errors are not an issue with the grader)')

--------------------
Sample Matrix:


array([[5, 2],
       [0, 9]])

--------------------
Numpy Solution:	44.99999999999999
Your Solution:	45.0

(Rounding errors are not an issue with the grader)


In [21]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 7

*15 Points*

With the 3x3 matrix, `A2`, create a function called `determinant3x3` that will calculate the determinant of any 3x3 matrix. See [this resource](https://www.mathsisfun.com/algebra/matrix-determinant.html) for a simple tutorial.

For this exercise, avoid using the `numpy.linalg.det()` numpy function. You may, however, use the functions created from Question 6 as part of your solution.

In [22]:
A3 = np.array([
    [7, 8, 1],
    [5, 5, 5],
    [7, 0, 3]
])

In [23]:
### GRADED

### YOUR SOLUTION HERE
def determinant3x3(M):
    pass

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*20)
print("Sample Matrix:")
A3 = matrix_maker(3,3)
print('-'*20)
print("Numpy Solution:\t{}".format(np.linalg.det(A3)))
print("Your Solution:\t{}".format(determinant3x3(A3)))
print()
print('(Rounding errors are not an issue with the grader)')

--------------------
Sample Matrix:


array([[0, 1, 6],
       [2, 1, 8],
       [1, 7, 5]])

--------------------
Numpy Solution:	76.0
Your Solution:	76.0

(Rounding errors are not an issue with the grader)


In [24]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

## Linear Independence

### Question 8

*10 Points*

For this question, refer to the lecture section on Linear Independence and the supplemental reading. Recall the definitions of dependence and independence:

**Dependence** means that at least one equation can be algebraically derived from the others (i.e., can be written as a linear combination of other equations). There are an **infinite** number of solutions that will satisfy the conditions of the equations! To know which solution you want, you have to feed in an x value. This makes the y value dependent on the x value. Attempting to solve a dependent system of linear equations via `numpy` will throw an error, because the solution depends on your inputs.

**Independence** means that no equation in the system can be algebraically derived from the others. In systems of linear equations with 2 equations, this means that the lines only intersect at one point, or the equations reflect parallel lines. There is either **no solution** or a **single solution**.

**Supplemental Reading**
* [Systems of Linear Equations](https://www.impan.pl/~pmh/teach/algebra/additional/lineareq.pdf)

<img src="images/linear_independence.png" width="500">

Consider the following sets of systems of equations

**System 1**

$$A1 = \begin{bmatrix}7 & 0 & 0 \\ 4 & 6 & 3 \\ 0 & 4 & 2\\ \end{bmatrix} ;   b1 = \begin{bmatrix}7 \\ 0 \\ 0 \end{bmatrix}$$


**System 2**

$$A2 = \begin{bmatrix}7 & 2 & 9 \\ 8 & 5 & 1 \\ 1 & 2 & 8\\ \end{bmatrix};    b2 = \begin{bmatrix}6 \\ 8 \\ 4 \end{bmatrix}$$


Your task for this question is as follows:
 - Determine which system of equations above is linearly independent. Indicate which system is linearly independent by assigning the string `"system 1"` or `"system 2"` to the variable `ans8a`.
 - For the system that is linearly independent, solve the system of equations and assign the solution to the variable `ans8b`. Round all values to 3 decimal places.
 
Note: There are no restrictions on using functions from Python libraries for this question.

In [25]:
# system of equations 1
A1 = np.array([
    [7, 0, 0],
    [4, 6, 3],
    [0, 4, 2]])
b1 = np.array([4, 8, 5])

# system of equations 2
A2 = np.array([
    [7, 2, 9],
    [8, 5, 1],
    [1, 2, 8]])
b2 = np.array([6, 8, 4])

In [26]:
### GRADED

### YOUR SOLUTION HERE
ans8a = ...
ans8b = ...


###
### YOUR CODE HERE
###

### For verifying answer:
print("Linear Independent System:\t'{}'".format(ans8a))
print("Solution for '{}':\t{}".format(ans8a, ans8b))

Linear Independent System:	'system 2'
Solution for 'system 2':	[0.30125523 1.07949791 0.19246862]


In [27]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

## Applied Linear Algebra

Eigendecomposition is a well defined linear algebra procedure that is useful for dimensionality reduction.

<img src="images/eigenvalue.svg" width="400">

For example, Principal Component Analysis (PCA) leverages eigendecomposition to reduce dimensionality by extracting new features (called components) that maximize the variance in the data.

The process for performing PCA is as follows:
1. Center the data around the origin by subtracting each value from its column mean
2. Compute the covariance matrix (square matrix)
3. Perform eigendecomposition on covariance matrix found in step 2.
4. Project data into subspace by taking the dot product of the transposed eigenvectors with the transposed mean centered data from step 1.

This process is demonstrated in the PCA lesson material.

[Back to top](#Index:) 

### Question 9

*15 Points*

For the following question, you will use the matrix `A` defined in the cell below. Do not round any values or answers.

Your task is as follows:
- Center the data around the origin by subtracting each value from it's column mean (computed below and bound to `col_means`). Assign this matrix to `ans9a`.
- Compute the covariance matrix (square matrix) from step 1. Assign this result to the variable `ans9b`.
- Perform eigendecomposition on covariance matrix found in step 2. Assign the vectors and values to the vaariables `ans9c` and `ans9d`, respectively.
- Project data into subspace by taking the dot product of the transposed eigenvectors with the transposed mean centered data from step 1. Assign the resulting matrix to the variable `ans9e`.

In [28]:
A, _ = make_regression(n_samples=1000, n_features = 5, noise=5, random_state=24)
display(A[:5])

array([[ 0.33971681, -0.7452978 ,  0.71999357, -0.10898928, -0.80510445],
       [-1.59009262,  0.34321702, -1.89241933,  1.7576691 ,  1.15028127],
       [ 0.26459075, -1.24608616,  0.74371496, -0.66519186, -0.97569435],
       [ 1.14975178,  0.92126388,  0.88806273, -0.24656481,  0.49593947],
       [-0.96989669,  0.85091215, -0.56442506, -0.6248215 , -0.88181707]])

In [29]:
### GRADED

### YOUR SOLUTION HERE
col_means = np.round(np.mean(A, axis=0), 3) # mean of each column
ans9a = ...
ans9b = ...
ans9c, ans9d = ..., ... # hint: np.linalg.eig()
ans9e = ...

###
### YOUR CODE HERE
###


In [30]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 10

*10 Points*

Now that you have implemented PCA from scratch, this question tasks you with implementing PCA using `scikit-learn`.

- Instantiate an instance of the class `PCA` with the default parameters. Bind this object to the variable `ans10a`.
- With the instantiated PCA object, fit and transform the matrix `A` (defined below). Bind the transformed data to the variable `ans10b`.
- What is the cumulative ratio of explained variance of the **first two components**? Assign the cumulative ratio of explained variance to the variable `ans10c` as a float. Round to 2 decimal places.

Hint: the `.explained_variance_ratio_` attribute may be helpful. See [docs](https://scikit-learn.org/stable/modules/generated/sklearn.decomposition.PCA.html#sklearn-decomposition-pca) for more.

In [31]:
from sklearn.decomposition import PCA
A, _ = make_regression(n_samples=1000, n_features = 5, noise=5, random_state=24)
display(A[:5])

array([[ 0.33971681, -0.7452978 ,  0.71999357, -0.10898928, -0.80510445],
       [-1.59009262,  0.34321702, -1.89241933,  1.7576691 ,  1.15028127],
       [ 0.26459075, -1.24608616,  0.74371496, -0.66519186, -0.97569435],
       [ 1.14975178,  0.92126388,  0.88806273, -0.24656481,  0.49593947],
       [-0.96989669,  0.85091215, -0.56442506, -0.6248215 , -0.88181707]])

In [32]:
### GRADED

### YOUR SOLUTION HERE
ans10a = ...
ans10b = ...
ans10c = ...

###
### YOUR CODE HERE
###

### For verifying answer:
print("PCA Instance:")
print()
print(ans10a)
print('-'*45)
print("Original Matrix:")
display(A[:5])
print('-'*45)
print("PCA Transformed Matrix:")
display(ans10b[:5])
print('-'*45)
print("Ratio of Explained Variance for Components 1 & 2: {}".format(ans10c))

PCA Instance:

PCA(copy=True, iterated_power='auto', n_components=None, random_state=None,
    svd_solver='auto', tol=0.0, whiten=False)
---------------------------------------------
Original Matrix:


array([[ 0.33971681, -0.7452978 ,  0.71999357, -0.10898928, -0.80510445],
       [-1.59009262,  0.34321702, -1.89241933,  1.7576691 ,  1.15028127],
       [ 0.26459075, -1.24608616,  0.74371496, -0.66519186, -0.97569435],
       [ 1.14975178,  0.92126388,  0.88806273, -0.24656481,  0.49593947],
       [-0.96989669,  0.85091215, -0.56442506, -0.6248215 , -0.88181707]])

---------------------------------------------
PCA Transformed Matrix:


array([[ 0.64076549,  0.75184713,  0.54150641, -0.66347072,  0.35965772],
       [-1.89895284, -1.22622239,  0.85221523,  1.88415338, -1.0866549 ],
       [ 0.52870039,  0.94079611,  0.59117183, -0.97437884,  1.0290358 ],
       [ 0.78541905,  0.1556773 , -1.29146843, -0.61474522, -0.83824043],
       [ 0.542303  , -0.53253399, -0.1541736 ,  1.24274883,  0.99817106]])

---------------------------------------------
Ratio of Explained Variance for Components 1 & 2: 0.44


In [33]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

### Question 11

*10 Points*

Another linear algebra technique for reducing dimensionality is Singular Value Decomposition (SVD).

Compute the $U$, $D$, and $V^T$ elements from performing SVD on the matrix $A$ below. Do not round any values.
- Assign the $U$ matrix to `ans11a`
- Assign the $D$ matrix to `ans11b`
- Assign the $V^T$ matrix to `ans11c`
- Using the 2 largest singular values from the diagonal ($D$), assign the reduced matrix (shape: 5x2) to the variable `ans11d`. This matrix can be computed by taking the dot product of $U$ and $D$ after $D$ has been cast into a diagonal matrix with the same number of rows and columns as A.

Hints:
- You may use `svd` function from `scipy.linalg` package
- The array, $D$, returned from the bullet above can be cast to the correct shape by running:
```
diagonal = np.empty((A.shape[0], A.shape[1]))
diagonal[:A.shape[0], :A.shape[0]] = diag(D)
```

In [34]:
A = np.array([
    [8, 7, 3, 1, 1, 9, 9, 5, 7, 6],
    [9, 8, 1, 4, 4, 8, 9, 5, 9, 3],
    [0, 4, 1, 8, 4, 9, 3, 4, 2, 7]
])

In [35]:
### GRADED

### YOUR SOLUTION HERE
ans11a = ...
ans11b = ...
ans11c = ...
ans11d = ...

###
### YOUR CODE HERE
###

### For verifying answer:
print('-'*40)
print("Original matrix")
display(A)
print('-'*40)
print("The 'U' matrix after decomposition")
display(ans11a)
print('-'*40)
print("The 'D' matrix after decomposition")
display(ans11b)
print('-'*40)
print("The 'VT' matrix after decomposition")
display(ans11c)
print('-'*40)
print("The reduced dimensionality matrix (U dot product with diagonal)")
display(ans11d)

----------------------------------------
Original matrix


array([[8, 7, 3, 1, 1, 9, 9, 5, 7, 6],
       [9, 8, 1, 4, 4, 8, 9, 5, 9, 3],
       [0, 4, 1, 8, 4, 9, 3, 4, 2, 7]])

----------------------------------------
The 'U' matrix after decomposition


array([[-0.62462913,  0.28752816,  0.72606198],
       [-0.65832617,  0.30622969, -0.68762637],
       [-0.42005368, -0.90749707, -0.00199267]])

----------------------------------------
The 'D' matrix after decomposition


array([[ 3.11147087e+01,  0.00000000e+00,  0.00000000e+00,
        -3.63790263e-01,  8.18629498e-02, -9.93154989e-02,
        -9.48833327e-02,  2.56948922e-02,  3.46591328e-01,
        -2.12708509e-01],
       [ 0.00000000e+00,  1.01700020e+01,  0.00000000e+00,
        -2.08214448e-01, -4.73218323e-01, -4.71441172e-01,
        -3.07756345e-01,  2.36470794e-01, -4.11598223e-01,
         2.57751123e-01],
       [ 0.00000000e+00,  0.00000000e+00,  4.29487636e+00,
         4.28900277e-02, -3.57947972e-01,  2.90439489e-01,
        -2.58491451e-01, -2.78425525e-01, -3.64662803e-01,
         5.30758954e-01]])

----------------------------------------
The 'VT' matrix after decomposition


array([[-0.35102269, -0.36379026, -0.09488333, -0.21270851, -0.15870785,
        -0.47144117, -0.41159822, -0.26016606, -0.35794797, -0.27842552],
       [ 0.49717714,  0.08186295,  0.02569489, -0.56514538, -0.20821445,
        -0.30775635,  0.25775112, -0.06501464,  0.29043949, -0.3646628 ],
       [-0.08851046, -0.0993155 ,  0.34659133, -0.47507418, -0.47321832,
         0.23647079,  0.0791507 ,  0.04289003, -0.25849145,  0.53075895],
       [ 0.3849956 , -0.61479651, -0.31244319,  0.39866133, -0.40632157,
        -0.02304054,  0.18671767, -0.01002794, -0.00405859,  0.12900938],
       [ 0.16374338, -0.5246444 , -0.02003731, -0.35163879,  0.72043745,
         0.05797699,  0.09153687,  0.00880935, -0.07190725,  0.19459271],
       [-0.16501812, -0.03607256, -0.55100088, -0.29081245, -0.10976548,
         0.67910319, -0.02950227, -0.12200088,  0.03178922, -0.30544916],
       [-0.49170874, -0.04645868, -0.02866331,  0.04129228,  0.02499184,
        -0.14974344,  0.83525739, -0.08201508

----------------------------------------
The reduced dimensionality matrix (U dot product with diagonal)


array([[-1.94351535e+01,  2.92416200e+00,  3.11834644e+00,
         1.98507297e-01, -4.47089992e-01,  1.37359811e-01,
        -2.16902538e-01, -1.50211954e-01, -5.99604919e-01,
         5.92338536e-01],
       [-2.04836271e+01,  3.11435656e+00, -2.95327025e+00,
         1.46238892e-01,  4.73284429e-02, -2.78701143e-01,
         1.45965590e-01,  2.46951491e-01, -1.03461978e-01,
        -1.46001229e-01],
       [-1.30698479e+01, -9.22924706e+00, -8.55827171e-03,
         3.41679976e-01,  3.95770681e-01,  4.68970574e-01,
         3.19659163e-01, -2.24834977e-01,  2.28663871e-01,
        -1.45617025e-01]])

In [36]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###


[Back to top](#Index:) 

## Normal Equations
One ubiquitous use of linear algebra is to the solve least squares regression problem. This approach involves using a system of linear equations to solve for the beta coefficents directly. For an intuitive understanding, the below walks you through arriving at the normal equation from a regression model purely from an algebraic standpoint. See [this post](https://math.stackexchange.com/questions/644834/least-squares-in-a-matrix-form) for a complete derivation of the normal equations.

Let's review the least squares regression line equation: 

<br>

$\hat y = \hat \beta_0 + \hat \beta_1x_1 + ... + \hat \beta_nx_n$

<br>

...which can be rewritten in matrix notation as:

<br>

$\hat y = X \hat \beta$

<br>

where:

- $\hat \beta$ are the least squares beta coeffients
- $X$ is a matrix of data
- $\hat y$ is a vector of least squares predictions
<br>


where $\hat y$ are our model's predictions.

<br>

Using this notation, we can apply the following linear algebra operations


**Start**

<br>

$\hat y = X\hat\beta$

<br>

**Multiply both sides by $X^T$**

<br>

$X^T\hat y = X^TX\hat\beta$


<br>

**Multiply both sides by $(X^TX)^{-1}$**

<br>

$(X^TX)^{-1}X^T\hat y = (X^TX)^{-1}(X^TX)\hat\beta$

<br>

**Since $(X^TX)^{-1}(X^TX)$ is equal to the identity matrix, the right side of the equation simplifies down to:**

<br>

$(X^TX)^{-1}X^T\hat y = \hat\beta$

<br>

This final equation is known as the Normal Equation. Let's demonstrate this calculation with sample data


Consider the sample dataset below.


| Obs | x1 | x2 | y |
|:-:|:-:|:-:|:-:|
|**1**|   0.96    |    0.10   |  61.5  |
|**2**|   1.45    |    1.05   | 113.4  |
|**3**|   0.16    |    0.51   |  21.9  |
|**4**|  -0.31    |   -0.99   | -42.0  |
|**5**|   1.32    |   -0.77   |  64.5  |
|**6**|  -1.07    |   -1.43   | -98.6  |
|**7**|   0.56    |    0.29   |  41.4  |
|**8**|  -1.62    |    0.21   | -94.9  |
|**9**|   0.67    |    1.88   |  84.7  |
|**10**|  -0.48    |    0.85   | -10.2  |


[Back to top](#Index:) 

### Question 12

*15 Points*

For this question, you will be asked to use the Normal Equation above to solve for the vector `b`.

Using the array `A` and vector `b`, compute the parameter vector `x` and assign the result to the variable `ans12`. Round all values to 3 decimal places.

In [37]:
# independent variables
A = np.array([
    [ 0.96,  0.1 ],
    [ 1.45,  1.06],
    [ 0.17,  0.52],
    [-0.32, -0.99],
    [ 1.33, -0.77],
    [-1.07, -1.44],
    [ 0.56,  0.3 ],
    [-1.63,  0.22],
    [ 0.68,  1.89],
    [-0.48,  0.85]
])

# dependent (target) variable
y = np.array([
    61.5,
    113.4,
    21.9,
    -42. ,
    64.2,
    -98.6,
    41.4,
    -95. ,  
    84.7, 
    -10.2
])

In [38]:
### GRADED

### YOUR SOLUTION HERE
ans12 = ...

###
### YOUR CODE HERE
###


In [39]:
###
### AUTOGRADER TEST - DO NOT REMOVE
###
